src/1_utils.js

```diff
@@ -553,3 +553,61 @@ utils.getBrowserLanguageCode = function() {
 	}
 	return code;
 };
+
+/**
+ * Returns the difference between two arrays. 
+ */
+utils.calculateDiffBetweenArrays = function(original, toCheck) {
+	var diff = [];
+	toCheck.forEach(function(element) {
+		if(original.indexOf(element) == -1) {
+			diff.push(element);
+		}
+	});
+
+	return diff;
+}
+
+/** 
+ * Validates the commerce-data object passed into branch.trackCommerceEvent().
+ * If there are invalid keys present then it will report back what those keys are.
+ * Note: The keys below are optional.
+ */
+utils.validateCommerceData = function(commerceData) {
+	var allowedInRoot = ['common', 'type', 'transaction_id', 'currency', 'revenue', 'revenue_in_usd', 'exchange_rate', 'shipping', 'tax', 'coupon', 'affiliation', 'persona', 'products'];
+	var allowedInProducts = ['sku', 'name', 'price', 'quantity', 'brand', 'category', 'variant'];
+
+	var invalidKeys = {
+		root: utils.calculateDiffBetweenArrays(allowedInRoot, Object.keys(commerceData)),
+		products: []
+	}
+
+	if(commerceData.hasOwnProperty('products')) {
+		commerceData["products"].forEach(function(product) {
+			utils.merge(invalidKeys["products"], utils.calculateDiffBetweenArrays(allowedInProducts, Object.keys(product)));
+		});
+	}
+
+	return invalidKeys;
+}
+
+/** 
+ * Builds an error message if partner provides invalid keys to commerce_data
+ */
+utils.buildErrorMessageForCommerceData = function(invalidKeys) {
+	var errMessage = 'Please remove the following keys from ';
+
+	if (invalidKeys['root'].length) {
+		errMessage += 'the root of commerce_data: ' + invalidKeys['root'].join(', ');
+	}
+
+	if (invalidKeys['root'].length && invalidKeys['products'].length) {
+		errMessage += '. In addition, please remove the following keys from commerce_data.products: ' + invalidKeys['products'].join(', ') + '.';
+	}
+
+	else if (invalidKeys['products'].length) {
+		errMessage += 'commerce_data.products: ' + invalidKeys['products'].join(', ');
+	}
+
+	return errMessage;
```

Comment:
Also, what's common with error messages is to not address every error in one pass. You could remove this helper function entirely and down in the original code just do 

```
if (invalidKeys["root"].length) {
	done('Please remove the following keys from the root of commerce_data: ' + invalidKeys['root'].join(', '));
}
else if (invalidKeys["products"].length) {
	done('Please remove the following keys from commerce_data.products: ' + invalidKeys['products'].join(', '));
}
```

You don't have to get it all in one pass, as nice as that is :) again, ultimately this is up to you and Gerrit, I'm definitely nitpicking.